### PR TITLE
RFC3164

### LINKED ISSUE #15: RFC 3164
Hi,

Any idea when that feature branch could be merged?

Thanks!

### PR DESCRIPTION
**Pull request by leodido**
This PR finally implements RFC3164 parsing! 😃 

It fixes #15 when merged in.

Notice it changes the `syslog.Message` interface and it introduces a `syslog.Base` struct containing the common fields between the two Syslog formats.

This means I'll need to release another major (v3) when completed 100% (see below).

Missing pieces at the moment:

- [ ] extensive test suite
- [x] option/mechanism to let the user choose the year of the timestamp
- [x] best-effort mode
- [x] option to choose the year for STAMP timestamps (`WithYear()` with strategies like `CurrentYear{}`, `Year{YYYY:2020}` etc.)
- [x] option to choose the time zone for STAMP timestamps (`WithTimezone()`)
- [x] option to let the parser ALSO match RFC3339 timestamps (`WithRFC3339()`)
- [x] examples
- [x] update README with examples referring to the new interfaces (#26)

### CODE REVIEW

**Comment by gianarb** [comment]
Hello @leodido !
Thank you for this contrib

```
--- FAIL: TestMachineParse (0.00s)
    --- FAIL: TestMachineParse/<34>Jan_12_06:30:00_xxx_apache_server:_1.2.3.4_-_- (0.00s)
        <autogenerated>:1: 
            	Error Trace:	machine_test.go:49
            	Error:      	Should NOT be empty, but was &{<nil> <nil> <nil> 0 <nil> <nil> <nil> <nil> <nil> <nil> <nil>}
            	Test:       	TestMachineParse/<34>Jan_12_06:30:00_xxx_apache_server:_1.2.3.4_-_-
FAIL
```
This is the failing test atm

**Comment by leodido** [comment]
This was the reason I put the PR as a draft (with WIP prefix). Anyway I have other commits to push :)

**Comment by leodido** [comment]
greeeeeeeny 💚

**Comment by leodido** [comment]
@goller I want your review my friend plzzz 🤗 

But please DO NOT merge yet :)

**Comment by goller** [comment]
@leodido you rock!

I think I'll start by reading through the structural changes.

Next, I'll read 3164 again and look closely at the grammar.

File: common/facility.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	6:  "line printer subsystem",
+	7:  "network news subsystem",
+	8:  "UUCP subsystem",
+	9:  "clock daemon",
+	10: "security/authorization messages",
+	11: "FTP daemon",
+	12: "NTP subsystem",
+	13: "log audit",
+	14: "log alert",
+	15: "clock daemon (note 2)", // (todo) > some sources reporting "scheduling daemon"
```
Comment (goller) [inline common/facility.go]: Should this (note 2) be removed?  The others with notes in the RFC that have notes are not specified here.

File: common/facility.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+var FacilityKeywords = map[uint8]string{
+	0:  "kern",
+	1:  "user",
+	2:  "mail",
+	3:  "daemon",
+	4:  "auth",
+	5:  "syslog",
+	6:  "lpr",
+	7:  "news",
+	8:  "uucp",
```
Comment (goller) [inline common/facility.go:43]: I'm seeing cron also defined as 9 in openbsd:
https://github.com/openbsd/src/blob/master/sys/sys/syslog.h#L98

File: common/facility.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	10: "authpriv",
+	11: "ftp",
```
Comment (goller) [inline common/facility.go:46]: [FreeBSD](https://github.com/freebsd/freebsd/blob/master/sys/sys/syslog.h#L107-L109) has 12, 13, and 14 defined as `ntp`, `security`, `console`, respectively.

Think those should be added?

File: common/functions.go [hunk 4]
```diff
@@ -40,3 +40,18 @@ func EscapeBytes(value string) string {
...
 }
+
+// InBetween tells whether value is into [min, max] range.
+func InBetween(val, min, max int) bool {
+	return val >= min && val <= max
+}
+
+// ValidPriority checks whether the given value is in the priority range [0, 191].
+func ValidPriority(priority uint8) bool {
+	return InBetween(int(priority), 0, 191)
```
Comment (goller) [inline common/functions.go:51]: Is this needed?  The grammar should forbid any numbers outside of this range.

Is it used for validation of a created and not parsed `SyslogMessage` struct?

File: common/common.rl [hunk 5]
```diff
@@ -16,14 +16,19 @@ bs = 0x5C;
...
 
+# 1..9
+nonzerodigit = '1'..'9';
+
 # 0..59
 sexagesimal = '0'..'5' . '0'..'9';
 
 # 01..31
-datemday = ('0' . '1'..'9' | '1'..'2' . '0'..'9' | '3' . '0'..'1');
+datemday = ('0' . nonzerodigit | '1'..'2' . '0'..'9' | '3' . '0'..'1');
```
Comment (goller) [inline common/common.rl]: I think the grammar may need to change.

The RFC says:

```
         If the day of the month is less
         than 10, then it MUST be represented as a space and then the
         number.  For example, the 7th day of August would be
         represented as "Aug  7", with two spaces between the "g" and
         the "7".

```

so, I think it should be

```
 datemday = (sp . nonzerodigit | '1'..'2' . '0'..'9' | '3' . '0'..'1');
```

Is that right?

File: rfc3164/machine.go.rl [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    m.pb = m.p
+}
+
+action set_prival {
+    output.priority = uint8(common.UnsafeUTF8DecimalCodePointsToInt(m.text()))
+	output.prioritySet = true
+}
+
+action set_timestamp {
+    if t, e := time.Parse(time.Stamp, string(m.text())); e != nil {
```
Comment (goller) [inline rfc3164/machine.go.rl]: The time.Time would be on year 0 with this type of parse:

https://play.golang.org/p/A6ueWMWKPE9

To get the time correctly, one would need to add the correct year.

In reading [Section 4.2](https://tools.ietf.org/html/rfc3164#section-4.2) I see:

```
         If the originally formed message has a TIMESTAMP in the HEADER
         part, then it SHOULD be the local time of the device within its
         timezone.
```

I don't know how one would know the device's timezone.  Perhaps the parser could have a [Location](https://golang.org/pkg/time/#Location) parameter?

File: rfc3164/machine.go.rl [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+	fhold;
+    fgoto fail;
+}
+
+pri = ('<' prival >mark %from(set_prival) $err(err_prival) '>') @err(err_pri);
+
+timestamp = (datemmm sp datemday sp hhmmss) >mark %set_timestamp @err(err_timestamp);
+
+# (todo) > RFC3164 says something about its maximum length?
+hostname = hostnamerange >mark %set_hostname $err(err_hostname);
```
Comment (goller) [inline rfc3164/machine.go.rl:137]: If I understand https://tools.ietf.org/html/std13 correctly, 255 octets is the max with hostname and domain.

However, RFC 3164 says `The Domain Name MUST NOT be included in the HOSTNAME field`.
I interpret this to mean that the max HOSTNAME is the max label in STD 13: 63 octets.

I think `hostnamerange` with 255 graphs is the right way to go.  Having extra space more safe I think.

File: rfc3164/machine.go.rl [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+# Section 4.1.3
+tag = (print - [ :\[]){1,32} >mark %set_tag @err(err_tag); # note(leodido) > alnum{1,32} is too restrictive (no dashes, see https://tools.ietf.org/html/rfc2234#section-2.1)
```
Comment (goller) [inline rfc3164/machine.go.rl]: In looking at `docs/rfc3164_tag.dot` the grammar allows

```
!..9
;..Z
\..~
```

The RFC says the following about tags:

```
   The TAG is a string of
   ABNF alphanumeric characters that MUST NOT exceed 32 characters.  Any
   non-alphanumeric character will terminate the TAG field and will be
   assumed to be the starting character of the CONTENT field
```

Should the tag grammar be further restricted?

I also see more information in [5.3](https://tools.ietf.org/html/rfc3164#section-5.3):

```
   Quite often, additional information is included at the beginning of
   the CONTENT field.  The format of "TAG[pid]:" - without the quote
   marks - is common.  The left square bracket is used to terminate the
   TAG field in this case and is then the first character in the CONTENT
   field.
```

File: rfc3164/machine.go.rl [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+content = '[' contentval ']'; # todo(leodido) > support ':' and ' ' too. Also they have to match?
+
+mex = visible+ >mark %set_message;
+
+msg = (tag content? ':' sp)? mex;
+
+fail := (any - [\n\r])* @err{ fgoto main; };
+
+main := pri timestamp sp hostname sp msg;
```
Comment (goller) [inline rfc3164/machine.go.rl]: Do you mind adding a comment that this library requires (for now) PRI, TIMESTAMP, and HOSTNAME?

I don't think this PR needs to handle sections [4.3.2](https://tools.ietf.org/html/rfc3164#section-4.3.2) and [4.3.3](https://tools.ietf.org/html/rfc3164#section-4.3.3).

File: rfc3164/machine.go.rl [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+visible = print | 0x80..0xFF;
+
+# The first not alphanumeric character starts the content part of the message part
+contentval = !alnum @err(err_contentstart) >mark print* %set_content @err(err_content);
+
+
+content = '[' contentval ']'; # todo(leodido) > support ':' and ' ' too. Also they have to match?
+
+mex = visible+ >mark %set_message;
```
Comment (goller) [inline rfc3164/machine.go.rl:152]: I think this is technically `content`

```
content = visible+ >mark %set_message;
```

and content is technically pid... maybe this?

```
 pidval = !alnum @err(err_contentstart) >mark print* %set_pidval @err(err_pidval);
 pid = '[' pidval ']'; # todo(leodido) > support ':' and ' ' too. Also they have to match?
```

File: rfc3164/machine.go.rl [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+action err_timestamp {
+    m.err = fmt.Errorf(errTimestamp, m.p)
+	fhold;
+    fgoto fail;
+}
+
+action err_hostname {
+	m.err = fmt.Errorf(errHostname, m.p)
```
Comment (goller) [inline rfc3164/machine.go.rl:106]: Does it matter to have a mix of tab and spaces?

File: rfc3164/machine_test.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+				Timestamp: syslogtesting.TimeParse(time.Stamp, "Jan 12 06:30:00"),
+				Hostname:  syslogtesting.StringAddress("xxx"),
+				Appname:   syslogtesting.StringAddress("apache"),
+				Message:   syslogtesting.StringAddress(`1.2.3.4 - - [12/Jan/2011:06:29:59 +0100] "GET /foo/bar.html HTTP/1.1" 301 96 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; fr; rv:1.9.2.12) Gecko/20101026 Firefox/3.6.12 ( .NET CLR 3.5.30729)" PID 18904 Time Taken 0`),
+			},
+		},
+		"",
+		nil,
+	},
+	// todo > other test cases pleaaaase
```
Comment (goller) [inline rfc3164/machine_test.go:148]: We could add these test cases as a follow-on PRs.  What do you think?

File: rfc3164/parser.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
+package rfc3164
+
+import (
+	"sync"
+
+	syslog "github.com/influxdata/go-syslog/v2"
+)
+
+// parser represent a RFC3162 parser with mutex capabilities.
```
Comment (goller) [inline rfc3164/parser.go]: Probably 3164 :)  I did take a minute and read a bit about [3162](https://tools.ietf.org/html/rfc3162) though

**Review by goller — CHANGES_REQUESTED** [review]
As usual, great stuff @leodido :)

I read the RFC while reviewing this PR and I spent most of my time checking the grammar.  The dot files really helped!

Regarding code structure changes (interface, base, builder), I think that's all just fine and great!  As you mentioned to me, this will require a new major release version.

Most of my suggestions were based on my interpretation of the RFC.  I think making PRI and HOSTNAME optional could be work in follow-on PRs.  Let me know what you think!

File: rfc3164/machine.go.rl (hunk 9 above)
Comment (goller) [reply rfc3164/machine.go.rl]: Do you imagine in the future something like this?

```
(pri (timestamp (sp hostname sp)?)?)? msg;
```

File: common/facility.go (hunk 1 above)
Comment (leodido) [reply common/facility.go]: I want to remove it too and I'll remove it.

Anyway I left `note 2` here in the past because there are two keys (9, 15) pointing to 2 clock daemons, as you also noted below.

File: common/facility.go (hunk 2 above)
Comment (leodido) [reply common/facility.go:43]: Yup, I confirm thanks for noticing it Chris :)

File: common/facility.go (hunk 3 above)
Comment (leodido) [reply common/facility.go:46]: Yes, and [RFC 5427](https://tools.ietf.org/html/rfc5427) - "Textual Conventions for Syslog Management" - too.

Thus I am going to use this RFC 5427 (fallbacking to RFC 5424 when needed).

File: common/functions.go (hunk 4 above)
Comment (leodido) [reply common/functions.go:51]: This is needed in 2 places:

- RFC 5424 builder
- RFC 5424 best-effort parsing mode uses it to understand whether the message is partial but still minimally correct or to reject completely (also in best-effort mode). This because such parsing option is technically an error that gets recovered if possible, except when the message was not even minimally indeed :)

File: rfc3164/machine.go.rl (hunk 6 above)
Comment (leodido) [reply rfc3164/machine.go.rl]: Yeah this was something I had to do but then I forgot. Thanks for reminding me.

My idea was to create an option for this parser - eg., `WithYear()` so that the user can decide the behavior.

WDYT?

File: rfc3164/machine.go.rl (hunk 11 above)
Comment (leodido) [reply rfc3164/machine.go.rl:106]: Gotta format it manually. Same old story 🤦‍♂ 😆 

Will do it at the end :)

File: rfc3164/machine.go.rl (hunk 8 above)
Comment (leodido) [reply rfc3164/machine.go.rl]: This was my concern too.

In fact, I initially used `alnum{1,32}` as you can see in the comment. Which is more restrictive than `print`.

Looking at various BSD-Syslog messages online, and looking at the interpretation of "ABNF alphanumeric" into [RFC 2234](https://tools.ietf.org/html/rfc2234#section-2.1), I decided to relax the grammar here using `print` (except starter characters for _content_ part).

The tag is conventionally used to store process names, and there are a lot of process names that contain dashes. Should we try using something in the middle? Like only alphanumeric plus dashes plus underscores and not all the print characters?

WDYT Chris? :)

File: rfc3164/machine.go.rl (hunk 10 above)
Comment (leodido) [reply rfc3164/machine.go.rl:152]: Chris this is not clear to me. Not getting it, sorry.

Why are you suggesting of renaming "mex" into "content" ?

File: common/common.rl (hunk 5 above)
Comment (leodido) [reply common/common.rl]: It is!

Good catch Chris, fixed :)

File: common/facility.go (hunk 1 above)
Comment (leodido) [reply common/facility.go]: Done

File: common/facility.go (hunk 2 above)
Comment (leodido) [reply common/facility.go:43]: Done, please take another look at it :)

File: common/facility.go (hunk 3 above)
Comment (leodido) [reply common/facility.go:46]: Should be 👌 now

File: rfc3164/machine_test.go (hunk 12 above)
Comment (leodido) [reply rfc3164/machine_test.go:148]: I'd prefer having 10 here covering most common messages. Maybe grabbing them from somewhere lese (we also have an issue here with some examples anyway).

Would you help me doing it? :)

File: rfc3164/parser.go (hunk 13 above)
Comment (leodido) [reply rfc3164/parser.go]: Tnx

File: rfc3164/machine.go.rl (hunk 9 above)
Comment (leodido) [reply rfc3164/machine.go.rl]: I agree that this PR does not need to handle 4.3.2 (valid PRI + no timestamp/invalid timestamp) and 4.3.3 (no or unidentifiable PRI).

Nevertheless, I think this parser should catch these situations - eg., throwing (non-generic) errors.
So that later, a user implementing a relay, or simply a wrapper around this parser, could detect those errors (for example no PRI, 4.3.3) and behave accordingly.

Clearly, this is easily doable as long I implement best-effort mode for this parser too. In this way when errors occur we can still recover the good parts of the message (useful in 4.3.2 use case for example).

Long story short: no, honestly, I do not imagine having **all** that pieces optional in the future.

File: rfc3164/machine_test.go (hunk 12 above)
Comment (goller) [reply rfc3164/machine_test.go:148]: @leodido Yup, I'll grab log messages.  Should I just add commits to this PR for the test cases?

File: rfc3164/machine.go.rl (hunk 6 above)
Comment (goller) [reply rfc3164/machine.go.rl]: I looked around at logstash and there have been a few attempts to make heuristics.
https://github.com/logstash-plugins/logstash-filter-date/pull/4#issue-26908778

Here is another heuristic:
https://github.com/logstash-plugins/logstash-filter-date/issues/51#issuecomment-309866094

This second link has three configurable heuristic strategies.  With your idea of `WithYear()` would be similar to `disambiguate=from_default`

I personally would use `nearest_default` to handle data coming in between Dec and Jan.

File: rfc3164/machine.go.rl (hunk 8 above)
Comment (goller) [reply rfc3164/machine.go.rl]: Good point.  I like alphanumeric/dashes/underscores.  I also think period, "."; for example, "howdy.sh"

I'm debating if we should just use what you have.  Ultimately, tag would be nearly any string until a semicolon or [.  

Now that I understand your reasoning, I think you should keep what you have already.  Do you mind adding a comment that "justifies" it?  Something like,
"while RFC3164 assumes only alphanumeric process names, many processes have additional characters."

File: rfc3164/machine.go.rl (hunk 9 above)
Comment (goller) [reply rfc3164/machine.go.rl]: Smart.  I like that point.  Use errors.

File: rfc3164/machine.go.rl (hunk 10 above)
Comment (goller) [reply rfc3164/machine.go.rl:152]: Yeah, I'm thinking mex into content, but, only because I was hoping to see pid/pidval.  Not that important to change I think.

File: rfc3164/machine.go.rl (hunk 7 above)
Comment (leodido) [reply rfc3164/machine.go.rl:137]: Oh, I see.

Anyway, I would leave this in [1,255]. At least in this PR.

An alternative (for a follow-up PR) would be to provide an option (eg. `WithoutDomainNameIntoHostname`?) to the parser to make it stricter ([1,63] octets).

WDYT?

File: rfc3164/machine.go.rl (hunk 8 above)
Comment (leodido) [reply rfc3164/machine.go.rl]: Agree!

File: rfc3164/machine.go.rl (hunk 7 above)
Comment (leodido) [reply rfc3164/machine.go.rl:137]: In the meantime, I insert a comment for future reference.

File: rfc3164/machine.go.rl (hunk 10 above)
Comment (leodido) [reply rfc3164/machine.go.rl:152]: Yeah, I just used tag and content as in the RFC3164.
Not that important imho :)

File: rfc3164/machine_test.go (hunk 12 above)
Comment (leodido) [reply rfc3164/machine_test.go:148]: Maybe we can add a bit of them and then merge it :)

File: rfc3164/machine.go.rl (hunk 7 above)
Comment (goller) [reply rfc3164/machine.go.rl:137]: Yeah, I could see giving more flexibility with those options in the future.  Good thinkin!

File: rfc3164/machine.go.rl (hunk 11 above)
Comment (leodido) [reply rfc3164/machine.go.rl:106]: Done

File: rfc3164/machine.go.rl (hunk 6 above)
Comment (leodido) [reply rfc3164/machine.go.rl]: These strategies are cool.

I just pushed a strategy pattern for year strategies and implemented a simple one as an example (current year) :)

This way other strategies can be easily implemented (and automatically used) later :)

File: rfc3164/machine.go.rl (hunk 8 above)
Comment (leodido) [reply rfc3164/machine.go.rl]: Commented the grammar ✅

**Comment by russorat** [comment]
Really looking forward to this!!!

**Comment by leodido** [comment]
Hey, @goller your test (https://github.com/influxdata/go-syslog/pull/27/commits/db2b9df413ba9a2f76f4d4cdec4b0e07d8eda753) was not passing because of missing whitespace at the start of the MSG part.

If you think that such BSD-syslog messages (no whitespace space between the TAG/CONTENT end - ':' - and the MSG start) are very common we could think to provide an option to parse them as correct ones.

**Comment by leodido** [comment]
Friendly ping! @goller 😃

**Review by fntlnz — APPROVED** [review]

**Comment by russorat** [comment]
is this really happening?! I'm so excited! Thank you for the great work here @leodido !

**Review by goller — APPROVED** [review]
LGTM... so many wonderful things :)
